### DISCUSSION TITLE
Copilot Usage Statistics

### DISCUSSION THREAD

**Discussion by danielnhr28** [community/community#54691] [Copilot Conversations]
### Select Topic Area

Question

### Body

is there a way to get copilot usage statistics? like how many times copilot has suggested code, how much time user has been using copilot?

thanks.

**Comment by ryanlua** [comment]
Currently there are no known ways to get GitHub copilot usage statistics from my knowledge and copilot does not record any data of such except for telemetry.

**Reply by danielnhr28** [reply]
when you mention telemetry what do you mean?

**Reply by ryanlua** [reply]
GitHub collecting data for analytics and such. Very common practice with most software.

**Comment by Akash1134** [comment]
Hi @danielnhr28 and @RyanLua , 

Currently, GitHub Copilot for Business shows usage information to organization and enterprise administrators. We plan to provide certain insights as part of a dashboard available to admins soon. The availability of this feature will be rolled out gradually to our customers. Thanks!

**Reply by danielnhr28** [reply]
hi @Akash1134 can i check what metrics are shown in the usage information? as from the github docs page i understand that we can only view how many users have access to GitHub Copilot across the organization, are there other metrics as well? 

thanks.

**Reply by zetaab** [reply]
at least we are interested of following:
- how many times copilot has suggested something
- how many times user has accepted the suggestion (pressed tab in case of vscode)
- how many times user has modified the suggested line after it was accepted

**Reply by RichardKCChen-SSSTC** [reply]
hi @Akash1134 We need to monitor member’s connection times,duration….,etc.Could I have API or any way to integrate it with power bi? Thanks

**Reply by Akash1134** [reply]
@RichardKCChen-SSSTC Currently, we do not have those tools available but our team is looking into all possible ways to provide you with the best options for usage and monitoring. Note that you can already view the [usage](https://docs.github.com/en/billing/managing-billing-for-github-copilot/viewing-your-github-copilot-usage#viewing-your-usage-for-github-copilot-for-business) for your organization. Currently you cannot extract analytics from GitHub Copilot. This is a capability anticipated for a later release.

**Reply by RichardKCChen-SSSTC** [reply]
hi @Akash1134 Thank for your reply. Now I only could got members’ last activity information by web export csv file.But my boss want more.Currently we have over 235 members use copilot in our org. We need to monitor when they access and how many times.Could you advise when will release new functions? And what information will provide?
Thanks

**Reply by johnpfeifferhelix** [reply]
Our organization is also excited by the promise of Copilot but want to ensure we can improve our understanding of usage:

- do some users have Licenses but are not using Copilot?
- is copilot enabled in an IDE and which IDE?
- which days (or times) are Copilot in use? (i.e. are devs coding with it in the morning and then focusing on PR reviews in the afternoon?)
- how many suggestions each user has received?

**Reply by Akash1134** [reply]
We've  released our beta version of the [User Management API](https://docs.github.com/en/rest/copilot/copilot-for-business?apiVersion=2022-11-28) for GitHub Copilot for Business.

Note: You must authenticate using an access token with the `manage_billing:copilot` scope and be an admin on the Copilot for Business organization you're updating to use this endpoint.

We'd love for you to try it out and let us know what you think!

**Reply by ensky** [reply]
Hi @Akash1134, a big thanks for the API! It's been of immense help to our organization for auditing purposes.

Expanding the API's capabilities to encompass more usages would be even more beneficial. Specifically:
- Determining the number of codes suggested by Copilot (Lines of Code from accepted suggestions / Total lines implemented).
  - Calculating this on a per-person basis is great, but calculating it per organization would also provide valuable insights.

By incorporating these usage statistics, we would gain profound insights into how Copilot is enhancing our organization's productivity!

**Reply by NickSchweitzer** [reply]
I don't see how the API is supposed to solve the request. I looked over the documentation, and it doesn't provide any more information regarding usage as the CSV available in the UI. Is there a timeframe for when more data will be available regarding usage as Number of Suggestions asked for per user? Maybe broken down by day?

**Reply by askpatrickw** [reply]
Really hard to justify $19 a dev with no proof the tool is even being used and to what extent.

**Reply by gagaus** [reply]
[Akash1134](https://github.com/Akash1134) is there a timeframe for when more data will be available regarding usage per user?

**Reply by devopscoefiba** [reply]
@Akash1134 We need insight data for the user, code, repo, etc. Probably the companies want to calculate how this tool affected their SDLC.  Thanks,

**Reply by ghost** [reply]
Thank you @Akash1134 for providing the API. However, as a company with dozens of developers, we require analytics to justify paying for this service. It is essential to add a minimum set of data to the API to track the number of code suggestions offered/accepted, interactions with the chat, and actions such as code refactor, explain code, generate tests, both on an individual developer basis and overall over a specified period.

**Reply by aghayants** [reply]
@Akash1134 do you have any updates on CoPilot Telemetry? How we can request this data, through email of admin tool?

**Reply by nick-mylo** [reply]
@Akash1134 echoing sentiments of @aghayants here, I'm keen to get some updates on the telemetry front for Copilot. Adding to the mix, I'm aligning with @askpatrickw's point: being an enterprise admin, it's part of my job to ensure that every dollar spent translates into tangible value for the team.

The heart of the matter is that without concrete metrics on usage (like how often copilot suggestions are being accepted or interacted with) broken down by user or even daily trends, it's challenging to paint a clear picture for stakeholders on the ROI of this tool.

Having this data at our disposal isn't just about oversight; it's about understanding how Copilot integrates into our development lifecycle and the efficiency gains we drive. These insights would be pivotal in making a case for the continued investment in the tool at $19 per dev.

appreciate all the efforts thus far and looking forward to those feature releases that can shed light on usage analytics!

**Reply by lrichter-ps** [reply]
@Akash1134 I'm giving a +1 on the last couple of comments. I'm planning to introduce Copilot to our engineering organization (around 70 developers), and of course, I need to make a good case for spending that money. A nice dashboard with the relevant statistics would be really helpful.

**Reply by ozagorodnii** [reply]
Yes, we definitely need the possibility to see stats by user for Business accounts. Maybe by a group if user-based stats are not available because of privacy or something.
Otherwise, we cannot gather proper metrics to justify further usage.

**Reply by satomic** [reply]
try this project, you will like it. https://github.com/satomic/copilot-usage-advanced-dashboard
![image](https://github.com/user-attachments/assets/e0cef464-b714-4b03-8324-fea760a9bea7)

**Comment by XSchelin** [comment]
Posting an update here that I found:

https://docs.github.com/en/enterprise-cloud@latest/early-access/copilot/copilot-usage-api

**Reply by askpatrickw** [reply]
That helps!

FYI, you can use the GH CLI to call that.. send to a file, etc..

gh api /orgs/exit83/copilot/usage

**Reply by zetaab** [reply]
For me these stats feels strange. The user count is much lower than what I can see in "github seats" pages used yesterday

**Reply by rmilkowski** [reply]
The way active users are presented (per day and in breakdowns only) makes it impossible to get total active users for a given period (entire week or all returned data). Also there should be a way to identify the active users and attribute stats to them as well.

**Reply by Noor-Hashem77** [reply]
Hmm this does not seem to work for me, im getting: 
![Screenshot 2024-02-22 at 3 17 24 PM](https://github.com/community/community/assets/97629236/bbfe36d0-e0b6-425e-abe9-58d831a71dc1)

**Reply by marcinkawcf** [reply]
@Noor-Hashem77 : probalby not enough permision or scope on the token. See here: https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28

**Reply by vagdwd** [reply]
https://github.com/github-copilot-resources/copilot-metrics-viewer

**Comment by freegor** [comment]
hi there,
We need the user's detailed statistic, to get company projects level adoption report, is it will be added?

**Comment by wangliguang111** [comment]
As a corporate account with many sub-accounts underneath, the company needs to know how everyone is using it and how much it is helping the company, without detailed statistics the company may not keep paying for it

**Comment by wangliguang111** [comment]
1, the stats for the chat function are not available
2, The statistics of detailed account is not available.
These statistics are really important for the company.
Because
1, it can evaluate whether people are using Copilot or not, and what kind of situation they are using.
2, what is the value to the company.

**Comment by wangliguang111** [comment]
Our company uses intellij idea. we bought a business account in november 2023, but so far there is no chat function, it's too much, it's not as good as some personal accounts that are suppose to be on the test list.

**Comment by riverjang-nsuslab** [comment]
Same. Still waiting for the update that would allow analysis of usage per user.

**Comment by ryan-vandusen** [comment]
In order to help with adoption, and sell the importance of AI tools to leadership, we 100% need some type of reporting to help show the value.  The use cases have been mentioned in this thread, so won't go through them, but this seems like a very basic ask.  We know it's increasing productivity, engineers say as much, but it would be great to see usability patterns.

**Reply by chrisRapsodo** [reply]
Second that. We have already started rolling out to a small group of volunteers within the organization and hope to roll it out more extensively. However, some quantitative data goes a long way to convincing leadership for this to happen.  Thanks.

**Comment by vpareek12** [comment]
I am in total agreement with the past comments. In addition to organization wide usage report. I'd also request you to provide some way for the individual to check his own usage report for example, how many lines of code adopted which were suggested by CoPilot, how many suggestions given, codes generated by copilot etc.

**Comment by vagdwd** [comment]
https://github.com/github-copilot-resources/copilot-metrics-viewer

**Comment by AdamSEMbyotic** [comment]
Has anyone tried using a third-party tool for Copilot measurement such as Opsera? https://www.opsera.io/solutions/opsera-github

They claim to provide dashboards for Copilot usage but I can't tell if they allow you to track individual developer usage.

**Comment by satomic** [comment]
just try this project, and you will like it. https://github.com/satomic/copilot-usage-advanced-dashboard

![image](https://github.com/user-attachments/assets/51203f1c-a51f-49f3-bb90-a9e1baf76d7a)

**Reply by froggogood** [reply]
This is a very good project, but I think I speak for everyone when I say we want these metrics per seat. But that is not your fault.

**Reply by satomic** [reply]
> This is a very good project, but I think I speak for everyone when I say we want these metrics per seat. But that is not your fault.

is [this](https://github.com/satomic/copilot-metrics-4-every-user) what you want? 🙂
